packages/transactions/src/builders.ts

```diff
@@ -1424,3 +1426,32 @@ export async function sponsorTransaction(
 
   return signer.transaction;
 }
+
+/**
+ * Estimates transaction byte length
+ * Context:
+ * 1) Multi-sig transaction byte length increases by adding signatures
+ *    which causes the incorrect fee estimation because the fee value is set while creating unsigned transaction
+ * 2) Single-sig transaction byte length remain same due to empty message signature which allocates the space for signature
+ * @param {transaction} - StacksTransaction object to be estimated
+ * @return {number} Estimated transaction byte length
+ */
+export function estimateTransactionByteLength(transaction: StacksTransaction): number {
+  const hashMode = transaction.auth.spendingCondition.hashMode;
+  // List of Multi-sig transaction hash modes
+  const multiSigHashModes = [AddressHashMode.SerializeP2SH, AddressHashMode.SerializeP2WSH];
+  // Check if its a Multi-sig transaction
+  if (multiSigHashModes.includes(hashMode)) {
+    const multiSigSpendingCondition: MultiSigSpendingCondition = transaction.auth
+      .spendingCondition as MultiSigSpendingCondition;
+    // Estimate signature bytes based on signaturesRequired value
+    // Formula: totalSignatureLength = signaturesRequired * (SIG_LEN_BYTES + 1 byte of type of signature)
+    const totalSignatureLength =
+      multiSigSpendingCondition.signaturesRequired * (RECOVERABLE_ECDSA_SIG_LENGTH_BYTES + 1);
```

Comment:
Great job on the explanatory comments here 🙏 
Could you also link or quote a SIP reference for this particular line? Assuming that matches up this new function looks good!

Reply:
> Also, is it possible to call this function with a partially signed multisig `StacksTransaction` object? For example, if 2 out of 3 required signatures are already added, does this function account for that?

This `estimateTransactionByteLength` function is getting called while making a unsigned transaction because fee value is set there. So the purpose of this function is to estimate the transaction byte length when the transaction is in unsigned stage. 
That's why i think checking if 2 out of 3 required signatures are already added is not required.

Reply:
> Great job on the explanatory comments here 🙏 Could you also link or quote a SIP reference for this particular line?

https://github.com/hirosystems/stacks.js/issues/1233#issue-1205643906

As mentioned by Aaron SIP reference might not have this information but it will require a closer look into `stack-node` implementation. 

I tried to show case the behaviour in test cases as well as in steps to reproduce script. 

🙏 if i am missing anything.

Reply:
> This `estimateTransactionByteLength` function is getting called while making a unsigned transaction because fee value is set there. So the purpose of this function is to estimate the transaction byte length when the transaction is in unsigned stage.

Ah I see. Is it still possible for devs to call this with a fully or partially signed multi-sig tx? If so, this function should either A) check that the input is unsigned and throw and error if not, or B) ensure the function works with unsigned, partial signed, and fully signed txs.

Reply:
+1  🙏

> [...] if not [...]

furthermore, we can additionally, add `Unsigned` to the function name to make it more obvious internally+externally.